### ISSUE TITLE
HTTP health check endpoint

### ISSUE THREAD

**Issue by bouk** [bensheldon/good_job#403]
Hi @bensheldon, thanks for the great gem.

We're using good_job in an environment that requires an HTTP health check to make sure the service is healthy. For Sidekiq there's a plugin available that does something similar: https://github.com/arturictus/sidekiq_alive

I couldn't figure out a good way to build a plugin like this for good_job, so I was wondering if you've dealt with this before and have any tips, or could suggest a good way to integrate an endpoint like this with good_job.

**Comment by bensheldon** [comment]
I am really excited by this challenge. Both because I think it's operationally important to have probes, and because it's an interesting technical challenge.

For the webserver part of the probe:

- If you're using GoodJob in `async` mode (i.e. running jobs inside the webserver), you can add the code below to your webserver's existing probe.
- For the CLI, it would need to have a tiny server built into it. That seems like it would be 20 lines of code to do with Rack... though I'm curious what the tiniest Rails API server implementation would be because the CLI is already initializing Rails probably YAGNI). 

For the content of the probes:

 - Readiness: the job schedulers have initialized and are ready to work jobs:
     ```ruby
     GoodJob::Scheduler.instances.all?(&:running?)
     ```
 - Liveliness: I have experienced Ops people arguing about what a Liveness probe should entail and whether it should check on service dependencies (e.g. database). This [sort of thing](https://srcco.de/posts/kubernetes-liveness-probes-are-dangerous.html). If I accept that liveliness includes a working database connection, then I would use:
 
    ```ruby
    GoodJob::Scheduler.instances.all?(&:running?) &&
      GoodJob::Notifier.instances.all?(&:listening?)
    ```
    
    
@bouk Is this a project you'd want to do? I think the smallest scope would be adding a simple Rack server (like [this](https://github.com/arturictus/sidekiq_alive/blob/master/lib/sidekiq_alive/server.rb)), starting the Rack server inside of a Concurrent::Future thread somewhere around [here](https://github.com/bensheldon/good_job/blob/f114b973f996da02ae0322608421ccc89668e7d4/lib/good_job/cli.rb#L95), and maybe making it conditional on some configuration like `good_job.config.probe_port`.

**Comment by bensheldon** [comment]
@aried3r Thanks! Yes, it should be `GoodJob::Scheduler.instances.any? && GoodJob::Scheduler.instances.all?(&:running?)`

Can you open an issue for what you're seeing about the instances not being started?

**Comment by bensheldon** [comment]
Linking a Reddit thread of another developer seeking similar functionality: https://reddit.com/r/ruby/comments/q4bbd3/ruby_webserver_started_in_background_thread/

**Comment by bensheldon** [comment]
@bouk I released this healthcheck feature in [v2.7.0](https://github.com/bensheldon/good_job/releases/tag/v2.7.0):

https://github.com/bensheldon/good_job#cli-http-health-check-probes
